# Part 0. Data Files

## 0.1 Download Data  

In [0]:
!pip install wget  

  Running setup.py bdist_wheel for wget: started
  Running setup.py bdist_wheel for wget: finished with status 'done'
  Stored in directory: C:\Users\Sukhvinder\AppData\Local\pip\Cache\wheels\40\15\30\7d8f7cea2902b4db79e3fea550d7d7b85ecb27ef992b618f3f
Successfully built wget


distributed 1.21.8 requires msgpack, which is not installed.
You are using pip version 10.0.1, however version 19.0.3 is available.
You should consider upgrading via the 'python -m pip install --upgrade pip' command.


In [0]:
import wget

link_to_data = 'https://github.com/tulip-lab/sit742/raw/master/Assessment/2019/data/wine.json'
DataSet = wget.download(link_to_data)

link_to_data = 'https://github.com/tulip-lab/sit742/raw/master/Assessment/2019/data/stopwords.txt'

DataSet = wget.download(link_to_data)

100% [................................................................................] 4158 / 4158

In [0]:
!ls

'ls' is not recognized as an internal or external command,
operable program or batch file.


## 0.2 Load Data  

In [0]:
import json
import pandas as pd
import matplotlib.pyplot as plt

In [0]:
file = 'wine.json'

In [0]:
w=pd.read_json(file)
w

country                                        description  \
0             Italy  Aromas include tropical fruit, broom, brimston...   
1          Portugal  This is ripe and fruity, a wine that is smooth...   
2                US  Tart and snappy, the flavors of lime flesh and...   
3                US  Pineapple rind, lemon pith and orange blossom ...   
4                US  Much like the regular bottling from 2012, this...   
5             Spain  Blackberry and raspberry aromas show a typical...   
6             Italy  Here's a bright, informal red that opens with ...   
7            France  This dry and restrained wine offers spice in p...   
8           Germany  Savory dried thyme notes accent sunnier flavor...   
9            France  This has great depth of flavor with its fresh ...   
10               US  Soft, supple plum envelopes an oaky structure ...   
11           France  This is a dry wine, very spicy, with a tight, ...   
12               US  Slightly reduced, this wine offers a chalky, t...   
13            Italy  This is dominated by oak and oak-driven aromas...   
14               US  Building on 150 years and six generations of w...   
15          Germany  Zesty orange peels and apple notes abound in t...   
16        Argentina  Baked plum, molasses, balsamic vinegar and che...   
17        Argentina  Raw black-cherry aromas are direct and simple ...   
18            Spain  Desiccated blackberry, leather, charred wood a...   
19               US  Red fruit aromas pervade on the nose, with cig...   
20               US  Ripe aromas of dark berries mingle with ample ...   
21               US  A sleek mix of tart berry, stem and herb, alon...   
22            Italy  Delicate aromas recall white flower and citrus...   
23               US  This wine from the Geneseo district offers aro...   
24            Italy  Aromas of prune, blackcurrant, toast and oak c...   
25               US  Oak and earth intermingle around robust aromas...   
26            Italy  Pretty aromas of yellow flower and stone fruit...   
27            Italy  Aromas recall ripe dark berry, toast and a whi...   
28            Italy  Aromas suggest mature berry, scorched earth, a...   
29               US  Clarksburg is becoming a haven for Chenin Blan...   
...             ...                                                ...   
129941           US  A Chardonnay with an unusual companion, 8% Sém...   
129942           US  This is classic in herbaceous aromas and flavo...   
129943        Italy  A blend of Nero d'Avola and Syrah, this convey...   
129944       Israel  Deep garnet in the glass, this has a nose of b...   
129945           US  Hailing from one of the more popular vineyards...   
129946      Germany  Plump, clingy peach and honey notes are cut wi...   
129947        Italy  A blend of 65% Cabernet Sauvignon, 30% Merlot ...   
129948    Argentina  Raspberry and cassis aromas are fresh and upri...   
129949           US  There's no bones about the use of oak in this ...   
129950           US  This opens with herbaceous dollops of thyme an...   
129951       France  Hugely spicy this rich wine is described as sw...   
129952           US  This Zinfandel from the eastern section of Nap...   
129953  New Zealand  Roughly two-thirds Cabernet and one-third Merl...   
129954  New Zealand  One of the more characterful Pinot Gris for th...   
129955  New Zealand  Like Dog Point's 2011 Chardonnay, this wine is...   
129956  New Zealand  The blend is 44% Merlot, 33% Cabernet Sauvigno...   
129957        Spain  Lightly baked berry aromas vie for attention w...   
129958  New Zealand  This blend of Cabernet Sauvignon-Merlot and Ca...   
129959       France  The granite soil of the Brand Grand Cru vineya...   
129960     Portugal  Fresh and fruity, this is full of red cherry f...   
129961        Italy  Intense aromas of wild cherry, baking spice, t...   
129962        Italy  Blackberry, cassis, grilled herb and toasted a...   
129963       Isr

In [0]:
w1=w.loc[:,'points':'price']
w1

points  price
0           87    NaN
1           87   15.0
2           87   14.0
3           87   13.0
4           87   65.0
5           87   15.0
6           87   16.0
7           87   24.0
8           87   12.0
9           87   27.0
10          87   19.0
11          87   30.0
12          87   34.0
13          87    NaN
14          87   12.0
15          87   24.0
16          87   30.0
17          87   13.0
18          87   28.0
19          87   32.0
20          87   23.0
21          87   20.0
22          87   19.0
23          87   22.0
24          87   35.0
25          87   69.0
26          87   13.0
27          87   10.0
28          87   17.0
29          86   16.0
...        ...    ...
129941      90   20.0
129942      90   35.0
129943      90   29.0
129944      90   25.0
129945      90   20.0
129946      90   17.0
129947      90   20.0
129948      90   43.0
129949      90   35.0
129950      90   35.0
129951      90   30.0
129952      90   22.0
129953      90   25.0
129954      90   15.0
129955      90   40.0
129956      90   19.0
129957      90   17.0
129958      90   35.0
129959      90   57.0
129960      90   48.0
129961      90   30.0
129962      90   40.0
129963      90   20.0
129964      90    NaN
129965      90   28.0
129966      90   28.0
129967      90   75.0
129968      90   30.0
129969      90   32.0
129970      90   21.0

[129971 rows x 2 columns]

In [0]:
w1=w1.dropna(subset=['points','price'])
w1

points  price
1           87   15.0
2           87   14.0
3           87   13.0
4           87   65.0
5           87   15.0
6           87   16.0
7           87   24.0
8           87   12.0
9           87   27.0
10          87   19.0
11          87   30.0
12          87   34.0
14          87   12.0
15          87   24.0
16          87   30.0
17          87   13.0
18          87   28.0
19          87   32.0
20          87   23.0
21          87   20.0
22          87   19.0
23          87   22.0
24          87   35.0
25          87   69.0
26          87   13.0
27          87   10.0
28          87   17.0
29          86   16.0
33          86   50.0
34          86   20.0
...        ...    ...
129940      91   36.0
129941      90   20.0
129942      90   35.0
129943      90   29.0
129944      90   25.0
129945      90   20.0
129946      90   17.0
129947      90   20.0
129948      90   43.0
129949      90   35.0
129950      90   35.0
129951      90   30.0
129952      90   22.0
129953      90   25.0
129954      90   15.0
129955      90   40.0
129956      90   19.0
129957      90   17.0
129958      90   35.0
129959      90   57.0
129960      90   48.0
129961      90   30.0
129962      90   40.0
129963      90   20.0
129965      90   28.0
129966      90   28.0
129967      90   75.0
129968      90   30.0
129969      90   32.0
129970      90   21.0

[120975 rows x 2 columns]

# Part 1: numeric anaysis

## 1.1 Explore the data distribution for each column.

In [0]:
w1.describe()

points          price
count  120975.000000  120975.000000
mean       88.421881      35.363389
std         3.044508      41.022218
min        80.000000       4.000000
25%        86.000000      17.000000
50%        88.000000      25.000000
75%        91.000000      42.000000
max       100.000000    3300.000000

## 1.2 Find the 10 varieties of wine which receives the highest number of  reviews

In [0]:
w['variety'].value_counts().head(10)

Pinot Noir                  13272
Chardonnay                  11753
Cabernet Sauvignon           9472
Red Blend                    8946
Bordeaux-style Red Blend     6915
Riesling                     5189
Sauvignon Blanc              4967
Syrah                        4142
Rosé                         3564
Merlot                       3102
Name: variety, dtype: int64

## 1.3 Find varieties of wine having the average price less than 20, with the average pointsat least 90

In [0]:
average=w.groupby(by='variety').mean()
average[(average['price']<20)&(average['points']>=90)]

points      price
variety                                    
Blauburgunder               93.0  19.000000
Caprettone                  92.0  19.000000
Kotsifali                   92.0  13.000000
Ondenc                      90.0  15.000000
Roussanne-Grenache Blanc    91.0  16.000000
Shiraz-Malbec               90.0  18.666667
Tinta Cao                   90.0  14.000000

## 1.4 Build statistic table 

In [0]:
winelistStatustable=w.groupby(['country','variety'])['variety'].size()
winelistStatustable=winelistStatustable.reset_index(level='country').groupby('country')['variety'].idxmax().reset_index()
                                                              
w1_list=w.groupby(by=['country','variety']).mean().round(2)
StatusTable=pd.merge(winelistStatustable,w1_list,how='inner',on=['country','variety'])
StatusTable.columns=['country','variety','avgpoint','avgprice']
StatusTable

country                   variety  avgpoint  avgprice
0                Argentina                    Malbec     87.50     27.96
1                  Armenia                   Kangoun     87.00     14.00
2                Australia                    Shiraz     89.64     47.29
3                  Austria          Grüner Veltliner     90.12     28.41
4   Bosnia and Herzegovina                   Blatina     88.00     12.00
5                   Brazil           Sparkling Blend     85.50     29.00
6                 Bulgaria                 Red Blend     89.18     22.00
7                   Canada                  Riesling     89.43     34.76
8                    Chile        Cabernet Sauvignon     86.69     20.02
9                    China            Cabernet Blend     89.00     18.00
10                 Croatia               Plavac Mali     87.59     34.00
11                  Cyprus               White Blend     87.00     14.20
12          Czech Republic          Cabernet Moravia     86.50     18.00
13                   Egypt                  Grenache     84.00       NaN
14                 England           Sparkling Blend     91.34     51.36
15                  France  Bordeaux-style Red Blend     88.82     40.90
16                 Georgia                  Saperavi     88.59     20.75
17                 Germany                  Riesling     90.09     44.26
18                  Greece               White Blend     86.96     19.68
19                 Hungary                   Furmint     89.92     48.24
20                   India                    Shiraz     91.50     16.00
21                  Israel        Cabernet Sauvignon     88.83     33.99
22                   Italy                 Red Blend     88.80     39.69
23                 Lebanon                 Red Blend     87.94     37.19
24              Luxembourg                  Riesling     88.00     23.00
25               Macedonia                Pinot Noir     86.00     15.00
26                  Mexico                 Red Blend     88.00     43.64
27                 Moldova                 Red Blend     87.82     22.55
28                 Morocco                 Red Blend     89.62     19.38
29             New Zealand           Sauvignon Blanc     88.05     18.72
30                    Peru                 Red Blend     84.50     28.75
31                Portugal            Portuguese Red     88.81     24.82
32                 Romania        Cabernet Sauvignon     86.12      9.94
33                  Serbia                 Red Blend     89.00     40.67
34                Slovakia                  Riesling     87.00     16.00
35                Slovenia               White Blend     89.10     31.50
36            South Africa              Chenin Blanc     87.85     19.88
37                   Spain               Tempranillo     87.54     31.02
38             Switzerland                Pinot Noir     89.00    160.00
39                  Turkey                 Red Blend     87.74     26.16
40                      US                Pinot Noir     89.61     44.59
41                 Ukraine           Sparkling Blend     82.83     11.00
42                 Uruguay                    Tannat     87.17     29.51

In [0]:
# save your table to 'statisticByState.csv'
winelistStatustable.to_csv('statisticByState.csv',sep=',')
w.loc[w1['points'].idxmax()]

country                                                          Australia
description              This wine contains some material over 100 year...
designation                                                           Rare
points                                                                 100
price                                                                  350
province                                                          Victoria
region_1                                                        Rutherglen
region_2                                                              None
taster_name                                                 Joe Czerwinski
taster_twitter_handle                                               @JoeCz
title                    Chambers Rosewood Vineyards NV Rare Muscat (Ru...
variety                                                             Muscat
winery                                         Chambers Rosewood Vineyards
Name: 345, dtype: object

## 1.5 Recommendations


Based on the analysis, which country/countries would you recommend *HOTEL TULIP* to source wine from? Pleas state your reasons.

---

Your Answer:

- From the above analysis I would recommend that tulip hotel should buy wine from Australia as it wines material contains the highest points that is 100 and the price is also not too hight 350 as a variety'Muscat' and i would recommend 'Chambers Rosewood Vineyards'
- ...

---


# Part 2. Text Analysis

## 2.1 extract high requency words in description

In [0]:
!pip install nltk

distributed 1.21.8 requires msgpack, which is not installed.
You are using pip version 10.0.1, however version 19.0.3 is available.
You should consider upgrading via the 'python -m pip install --upgrade pip' command.


In [0]:
import re
import nltk
from nltk.tokenize import RegexpTokenizer
from nltk.probability import *
from itertools import chain
#from tqdm import tqdm
import codecs

In [0]:
with open('stopwords.txt') as f:
    stop_words = f.read().splitlines()
stop_words = set(stop_words)
stop_words

{'a',
 "a's",
 'able',
 'about',
 'above',
 'according',
 'accordingly',
 'across',
 'actually',
 'after',
 'afterwards',
 'again',
 'against',
 "ain't",
 'all',
 'allow',
 'allows',
 'almost',
 'alone',
 'along',
 'already',
 'also',
 'although',
 'always',
 'am',
 'among',
 'amongst',
 'an',
 'and',
 'another',
 'any',
 'anybody',
 'anyhow',
 'anyone',
 'anything',
 'anyway',
 'anyways',
 'anywhere',
 'apart',
 'appear',
 'appreciate',
 'appropriate',
 'are',
 "aren't",
 'around',
 'as',
 'aside',
 'ask',
 'asking',
 'associated',
 'at',
 'available',
 'away',
 'awfully',
 'b',
 'be',
 'became',
 'because',
 'become',
 'becomes',
 'becoming',
 'been',
 'before',
 'beforehand',
 'behind',
 'being',
 'believe',
 'below',
 'beside',
 'besides',
 'best',
 'better',
 'between',
 'beyond',
 'both',
 'brief',
 'but',
 'by',
 'c',
 "c'mon",
 "c's",
 'came',
 'can',
 "can't",
 'cannot',
 'cant',
 'cause',
 'causes',
 'certain',
 'certainly',
 'changes',
 'clearly',
 'co',
 'com',
 'come',
 'c

In [0]:
tokenizer = RegexpTokenizer(r"\w+(?:[-']\w+)?")
sentencelist=[]
for words in w["description"]:
    sentence=[a for a in tokenizer.tokenize(words) if a not in stop_words]
    sentencelist.append(sentence)
tokenizedwords=[word for word in sentence for sentence in sentencelist]
tokenizedwords

['Big',
 'Big',
 'Big',
 'Big',
 'Big',
 'Big',
 'Big',
 'Big',
 'Big',
 'Big',
 'Big',
 'Big',
 'Big',
 'Big',
 'Big',
 'Big',
 'Big',
 'Big',
 'Big',
 'Big',
 'Big',
 'Big',
 'Big',
 'Big',
 'Big',
 'Big',
 'Big',
 'Big',
 'Big',
 'Big',
 'Big',
 'Big',
 'Big',
 'Big',
 'Big',
 'Big',
 'Big',
 'Big',
 'Big',
 'Big',
 'Big',
 'Big',
 'Big',
 'Big',
 'Big',
 'Big',
 'Big',
 'Big',
 'Big',
 'Big',
 'Big',
 'Big',
 'Big',
 'Big',
 'Big',
 'Big',
 'Big',
 'Big',
 'Big',
 'Big',
 'Big',
 'Big',
 'Big',
 'Big',
 'Big',
 'Big',
 'Big',
 'Big',
 'Big',
 'Big',
 'Big',
 'Big',
 'Big',
 'Big',
 'Big',
 'Big',
 'Big',
 'Big',
 'Big',
 'Big',
 'Big',
 'Big',
 'Big',
 'Big',
 'Big',
 'Big',
 'Big',
 'Big',
 'Big',
 'Big',
 'Big',
 'Big',
 'Big',
 'Big',
 'Big',
 'Big',
 'Big',
 'Big',
 'Big',
 'Big',
 'Big',
 'Big',
 'Big',
 'Big',
 'Big',
 'Big',
 'Big',
 'Big',
 'Big',
 'Big',
 'Big',
 'Big',
 'Big',
 'Big',
 'Big',
 'Big',
 'Big',
 'Big',
 'Big',
 'Big',
 'Big',
 'Big',
 'Big',
 'Big',
 'Big',


In [0]:
from itertools import takewhile
fdist = FreqDist(tokenizedwords)
common_words=fdist.most_common()
high_freq_comm=dict(takewhile(lambda x:x[1]>5000,common_words))
key_word=high_freq_comm.keys()
sorted_high_common={}
for key in sorted(key_word):
    sorted_high_common[key]=high_freq_comm[key]
sorted_high_common

{'Big': 129971,
 'Drink': 129971,
 'Lychees': 129971,
 'aftertaste': 129971,
 'dominate': 129971,
 'feel': 129971,
 'fruit': 129971,
 'giving': 129971,
 'intense': 129971,
 'off-dry': 129971,
 'opulent': 129971,
 'powered': 129971,
 'profile': 129971,
 'rich': 129971,
 'rounded': 129971,
 'spiciness': 129971,
 'texture': 129971}

In [0]:
with open('HighFreq.txt','a')as file:
    file.write(str(sorted_high_common))

## 2.2 Find key words for describing Shiraz using TF-IDF

In [0]:
shirazsentences=w[w['variety']=='Shiraz']['description']
shirazsentences

232       Lifted cedar and pine notes interspersed with ...
293       This wine displays ample concentration in its ...
356       Dusty, firm, powerful: just a few apt descript...
365       The Taylor family selected Clare Valley for it...
628       This plush, full-bodied wine is already delici...
636       The Torbreck wines, with the emphasis they pla...
893       This is an unusually complex Shiraz for the pr...
997       Blended from a patchwork of old vineyards thro...
1008      Intense and focused, this good value Coonawarr...
1074      Sourced from vines planted in 1847, this is a ...
1154      This wine boasts admirable complexity, but com...
1715      Raspberry and cherry mingle with essence of mo...
1723      Tart blackberry and vanilla flavors mark this ...
1886      Big, bold and structured, this is a dry yet mo...
1934      Mint marks the nose of this otherwise darkly a...
2256      Not a wine for the faint of heart. Alluring no...
2594      Made in a superripe, dense, al

In [0]:
from sklearn.feature_extraction.text import TfidfVectorizer
tf_idf = TfidfVectorizer(analyzer='word', stop_words = 'english')
x=tf_idf.fit_transform(shirazsentences)
idf=tf_idf.idf_
tfids_values=dict(zip(tf_idf.get_feature_names(),idf))
tfids_values

{'01': 6.631211781821365,
 '10': 5.0907667408742165,
 '100': 5.783913921434162,
 '11': 6.343529709369585,
 '12': 6.631211781821365,
 '120': 7.03667688992953,
 '125': 7.03667688992953,
 '128': 6.631211781821365,
 '13': 6.343529709369585,
 '14': 5.783913921434162,
 '15': 4.638781617131159,
 '16': 5.650382528809639,
 '171': 7.03667688992953,
 '18': 6.631211781821365,
 '1843': 7.03667688992953,
 '1847': 7.03667688992953,
 '1892': 7.03667688992953,
 '1893': 7.03667688992953,
 '1912': 7.03667688992953,
 '1952': 7.03667688992953,
 '1971': 7.03667688992953,
 '1973': 7.03667688992953,
 '20': 6.120386158055375,
 '2002': 7.03667688992953,
 '2003': 6.343529709369585,
 '2004': 6.120386158055375,
 '2005': 5.244917420701475,
 '2006': 5.4272389774954295,
 '2007': 4.594329854560326,
 '2008': 5.244917420701475,
 '2009': 5.650382528809639,
 '2010': 4.397619560314272,
 '2010s': 7.03667688992953,
 '2011': 5.93806460126142,
 '2012': 4.896610726433259,
 '2013': 5.4272389774954295,
 '2014': 5.650382528809639,

In [0]:
score=list((key,value.round(2)) for (key,value) in tfids_values.items() if value>0.4)
score.sort()
score

[('01', 6.63),
 ('10', 5.09),
 ('100', 5.78),
 ('11', 6.34),
 ('12', 6.63),
 ('120', 7.04),
 ('125', 7.04),
 ('128', 6.63),
 ('13', 6.34),
 ('14', 5.78),
 ('15', 4.64),
 ('16', 5.65),
 ('171', 7.04),
 ('18', 6.63),
 ('1843', 7.04),
 ('1847', 7.04),
 ('1892', 7.04),
 ('1893', 7.04),
 ('1912', 7.04),
 ('1952', 7.04),
 ('1971', 7.04),
 ('1973', 7.04),
 ('20', 6.12),
 ('2002', 7.04),
 ('2003', 6.34),
 ('2004', 6.12),
 ('2005', 5.24),
 ('2006', 5.43),
 ('2007', 4.59),
 ('2008', 5.24),
 ('2009', 5.65),
 ('2010', 4.4),
 ('2010s', 7.04),
 ('2011', 5.94),
 ('2012', 4.9),
 ('2013', 5.43),
 ('2014', 5.65),
 ('2014s', 7.04),
 ('2015', 4.09),
 ('2016', 5.02),
 ('2017', 4.79),
 ('2018', 4.15),
 ('2019', 6.34),
 ('2020', 3.19),
 ('2021', 6.12),
 ('2022', 5.43),
 ('2023', 5.78),
 ('2025', 3.29),
 ('2027', 6.63),
 ('2028', 7.04),
 ('2030', 4.15),
 ('2032', 6.63),
 ('2037', 6.63),
 ('2040', 7.04),
 ('21', 7.04),
 ('25', 6.63),
 ('28', 7.04),
 ('30', 7.04),
 ('300', 7.04),
 ('34th', 7.04),
 ('40', 7.04),

In [0]:
with open("ShirazKey.txt",'w') as file:
    file.write(str(score))